### ISSUE TITLE
Max margin objectives

### ISSUE THREAD

**Issue by vzhong** [fchollet/keras#150]
Hi, 

Thanks for developing Keras - it's really awesome! I'm looking to use Keras for my research and have a question regarding implementing objectives that require negative sampling. For example, suppose I want to optimize the max margin loss as follows:

```python
Ppos = model.predict(Xpos)
Pnet = model.predct(Xneg)
loss = T.sum(T.maximum(0, 1 - Ppos + Pneg))
```

Is there a way to do this in Keras?

Another example would be negative sampling, where we'd want to optimize something like:

```python
loss = -T.log(Ppos)
for Pneg in Pnegs:
    loss -= T.log(1-Pneg)
```

**Comment by fchollet** [comment]
That would be possible, batch-wise. You can implement a new loss function, assuming binary classes `{0, 1}`, such as:

```python
def max_margin(y_true, y_pred):
    return T.sum(T.maximum(0., 1. - y_pred*y_true + y_pred*(1. - y_true)))
```
Which returns the max margin loss on the current batch. Am I correctly interpreting what you are trying to do?

**Comment by vzhong** [comment]
Sorry I probably wasn't very clear :)

I'd like to use this for triplet predictions, where given some triplet `e1, r, e2`, `Xpos` corresponds to the concatenated embeddings of subject `e1`, relation `r`, and object `e2`. For each triplet, we generate a corresponding corrupt triplet `e1, r, ec`, where `ec` is some randomly sampled object entity.  `Xneg` corresponds to the concatenated embeddings of `e1, r, ec`.

Similarly, `Ppos` corresponds to the model's scores for the correct triplet and `Pneg` corresponds to the model's scores for the corrupted triplet. The quantity `margin = 1 - Ppos + Pneg` is such that `margin[i]` gives how much the score for the corrupted example `i` exceeds the score of the original example `i`.

I suppose I can hack a solution to this, which is to reserve the first half of the input features `X` for the positive examples, and the second half of the input features `X` for the corresponding corrupted examples, where `X[n/2+i]` is the corrupted example for `X[i]`.

In this case the max margin loss would be:

```python
def max_margin(y_true, y_pred):
    num_ex = y_pred.shape[0]
    y_pos = y_pred[:num_ex]
    y_neg = y_pred[num_ex:]
    return T.sum(T.maximum(0., 1. - y_pos + y_neg)))
```

Similarly, suppose we want to do negative sampling with 5 negative samples per positive sample, we can reserve the first 6th of `X` for the positives, the next 5 6th of `X` for each of the negatives. But this is rather hacky. Is there cleaner way of constructing objectives that don't fit cleanly into the `y_true`, `y_pred` construction?

**Comment by fchollet** [comment]
It's not all that hacky: it gets the job done and it's very simple to put together and manage. It does by-pass the implicit assumption that loss should be computable on a per-sample basis, but that shouldn't matter. 

Our example of word embedding with negative sampling works essentially on that principle.

**Comment by vzhong** [comment]
Ok thanks Francois :+1:

**Comment by eshijia** [comment]
Hi, I am doing something about triplet predictions. @fchollet Could you show the example of word embedding with negative sampling explicitly or tell me where is the code? I haven' t found any code related to that in the "example" folder. Thanks!

**Comment by stephantul** [comment]
@eshijia The example he is referring to can be found [here](https://github.com/fchollet/keras/blob/1aaab7a228fdd6f3ef53e888e08bc2afd6b4213b/examples/skipgram_word_embeddings.py).

**Comment by eshijia** [comment]
@stephantul Thanks for your reply. That's cool! I know what I should learn now.

**Comment by nooralahzadeh** [comment]
I have a dataset with positive and negative samples. how can I train and test with max margin objective in keras? the dataset is as following:
<sentA, CategoryA ,+>
<sentA, CategoryB,+>
<sentA, CategoryC,->
and so on
Do i need to update  per sample? 
Thanks
